client-implementation.md

```diff
@@ -0,0 +1,67 @@
+# Client Implementations
+
+This document describes client implementation recommendations for interoperability with registry servers and other clients using the OCI Distribution Spec.
+
+## Backwards Compatibility
+
+Client implementations SHOULD support registries that implement partial or older versions of the OCI Distribution Spec.
+This section describes how client fallback procedures when an API is not available.
+
+### Referrers API
+
+The Referrers API here is described by [Listing Referrers](spec.md#listing-referrers) and [end-12a](spec.md#endpoints).
+
+A client that pushes an Image or Artifact manifest with a defined `Refers` field MUST verify the Referrers API is available.
```

Comment:
```suggestion
Before pushing an Image or Artifact manifest with a defined `Refers` field a client MUST first verify the Referrers API is available.
```

Reply:
** how will they do that? is there a known nil/empty set request?

Reply:
After the artifact is pushed, there will be content in the API. Before that, there would be an Index with an empty manifest list.

Reply:
ok, I've now skimmed both of the spec change PRs.. :-)

I believe when you say there would be an index with an empty manifest list you are specifically talking about the image.index that is used for "information about a set of images that can span a variety of architectures and operating systems" (where each image is defined by an image manifest) for which it seems this new proposal is now also using for the new artifact manifest. Is there some expression here that is switching to using a more generalized term "index" when referring to the image.index which is now for image manifests and/or artifact manifests?   

If so let's make sure that's described somewhere, then make sure to go through all the existing image index terms and see where it needs to be index.

Reply:
> Before pushing an Image or Artifact manifest with a defined `Refers` field a client MUST first verify the Referrers API is available.

I don't believe order is important, you can push the manifest and then afterwords verify that the Referrers API is available. In fact to push the change to the tag schema Index, you would need to first push the manifest so the Index contains valid descriptors.

"index with an empty manifest list" is the response from the Referrers API when the API exists but no manifest with a matching refers field is found.

> I believe ... you are specifically talking about the image.index that is used for "information about a set of images that can span a variety of architectures and operating systems"

It's a reference to https://github.com/opencontainers/image-spec/blob/main/image-index.md. Each descriptor in there points to a manifest, which isn't limited to an image manifest in the spec today. We can update the image-spec to clarify that the Index MUST support the new Artifact manifest as a descriptor.

Reply:
kk 

interesting, the overloading of the image.index manifest schema for the referrers api, .. need to think it over, what needs to be said about it in the spec wrt client use of a partial index etc.

yes (IMO) to the make sure the image spec (image.index) includes MUST language for supporting artifact manifest type.. (for the next version)